File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
  const tTally = {};
  const window = {};
  for (let i = 0; i < t.length; i++) {
    tTally[t[i]] = tTally[t[i]] || 0;
    tTally[t[i]]++;
  }
  let [have, lb, ub, l] = [0, -1, s.length + 1, 0];
  const need = Object.keys(tTally).length;
  
  for (let r = 0; r < s.length; r++) {
    if (tTally[s[r]] !== undefined) {
      window[s[r]] = window[s[r]] || 0;
      window[s[r]]++;
      if (window[s[r]] === tTally[s[r]]) have++;
      while (have === need) {
        if (r + 1 - l < ub - lb) [lb, ub] = [l, r + 1];
        if (tTally[s[l]] !== undefined) {
          window[s[l]]--;
          if (window[s[l]] < tTally[s[l]]) have--;
        }
        l++;
      }
    }
  }

  return lb < 0 ? "" : s.slice(lb, ub);
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.go
func minWindow(s string, t string) string {
  tTally, window := make(map[byte]int), make(map[byte]int)
  for i:=0;i<len(t);i++ {
    tTally[t[i]]++
  }
  have, need, lb, ub, l := 0, len(tTally), -1, len(s) + 1, 0
  for r:=0;r<len(s);r++ {
    if _, ok := tTally[s[r]]; ok {
      window[s[r]]++
      if window[s[r]] == tTally[s[r]] {
        have++
      }
      for have == need {
        if r + 1 - l < ub - lb {
          lb, ub = l, r + 1
        }
        if _, ok = tTally[s[l]]; ok {
          window[s[l]]--
          if window[s[l]] < tTally[s[l]] {
            have--
          }
        }
        l++
      }
    }
  }
  if lb < 0 {
    return ""
  } else {
    return s[lb:ub]
  }
}